src/Graphics/ColorBufferReader.cpp

```diff
@@ -0,0 +1,89 @@
+
+#include "ColorBufferReader.h"
+#include "FramebufferTextureFormats.h"
+#include "Context.h"
+#include "Parameters.h"
+#include <algorithm>
+#include <cstring>
+
+namespace graphics {
+
+	ColorBufferReader::ColorBufferReader(CachedTexture * _pTexture)
+		: m_pTexture(_pTexture)
+	{
+		m_pixelData.resize(m_pTexture->textureBytes);
+		m_tempPixelData.resize(m_pTexture->textureBytes);
+	}
+
+	const u8* ColorBufferReader::_convertFloatTextureBuffer(const u8* _gpuData, u32 _width, u32 _height,
+		u32 _heightOffset, u32 _stride)
+	{
+		std::copy_n(_gpuData, m_tempPixelData.size(), m_tempPixelData.data());
+		u8* pixelDataAlloc = m_pixelData.data();
+		float* pixelData = reinterpret_cast<float*>(m_tempPixelData.data());
+		const int colorsPerPixel = 4;
+		const int widthPixels = _width * colorsPerPixel;
+		const int stridePixels = _stride * colorsPerPixel;
+
+		for (u32 index = 0; index < _height; ++index) {
+			for (u32 widthIndex = 0; widthIndex < widthPixels; ++widthIndex) {
+				u8& dest = *(pixelDataAlloc + index*widthPixels + widthIndex);
+				float& src = *(pixelData + (index+_heightOffset)*stridePixels + widthIndex);
+				dest = static_cast<u8>(src*255.0);
+			}
+		}
+
+		return pixelDataAlloc;
+	}
+
+	const u8* ColorBufferReader::_convertIntegerTextureBuffer(const u8* _gpuData, u32 _width, u32 _height,
+		u32 _heightOffset, u32 _stride)
+	{
+		const int colorsPerPixel = 4;
+		const size_t widthBytes = _width * colorsPerPixel;
+		const int strideBytes = _stride * colorsPerPixel;
```

Comment:
What I'm finding, at least for mobile, that it seems faster to do one big copy instead of many small copies from GPU memory. I didn't test the performance difference for writing to GPU memory, but I assumed that it would be similar.

What do you think?